packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts

```diff
@@ -482,8 +518,46 @@ export async function visualizeChildData(
  * Determines whether or not the provided value is an {@link ISharedObject}, for the purposes of this library.
  * @remarks Implemented by checking for the particular properties / methods we use in this module.
  */
-function isSharedObject(value: Partial<ISharedObject>): value is ISharedObject {
+function isSharedObject(value: unknown): value is ISharedObject {
+	return (
+		(value as SharedObject).id !== undefined &&
+		(value as SharedObject).attributes?.type !== undefined &&
+		(value as SharedObject).on !== undefined
+	);
+}
+
+/**
+ * Determines whether or not the provided value is an {@link DataObject}, for the purposes of this library.
+ * @remarks Implemented by checking for the particular properties / methods we use in this module.
+ * @remarks `value` is type casted as a workaround to access the `root` property of the {@link DataObject} (private).
+ */
+function isDataObject(value: unknown): value is DataObject {
+	if (
+		(value as DataObject).initializeInternal === undefined ||
+		(value as { getUninitializedErrorString(): string }).getUninitializedErrorString ===
+			undefined
+	) {
+		return false;
+	}
+
+	// If root is missing, throw an error instead of returning false
+	const root = (value as { readonly root?: ISharedDirectory }).root;
+	if (!root) {
+		throw new Error("DataObject must have a `root` property, but it was undefined.");
```

Comment:
@alexvy86 @Josmithr @chentong7 

Error case included to prevent `undefined` root.